tests/unit/test_environment_component.py

```
@@ -0,0 +1,179 @@
+        "pulumi_app.environment.pulumi.get_project", return_value="project-fallback"
+    ):
+        mock_config.return_value.get.return_value = None
+        outputs = _run_pulumi_program(program)
+
+    assert outputs["serviceName"] == "project-fallback"
+    assert outputs["stackTag"] == "project-fallback-qa"
+    assert outputs["defaultTags"] == {"Project": "project-fallback", "Environment": "qa"}
+
+def test_environment_prefers_explicit_over_config_and_default() -> None:
+    """Covers mutation that swaps fallback order for environment."""
+    def program():
```

Ruff (0.14.4)

Missing return type annotation for private function `program`

Add return type annotation: `None`

(ANN202)